### PR TITLE
Add AER (autoencoder w/ regression) model to benchmark

### LINKED ISSUE #286: Add ``AER`` Model
Add a new model, namely ``AER``, an AutoEncoder Regressor model.

**Model Architecture**
image

**Description**
Prediction-based and reconstruction-based anomaly scoring methods each have their own success and limitations. AER is a model that can predict both prediction-based and reconstruction-based anomaly scores and combine those scores to leverage the success of both scoring methods. Assuming the input length is **n**, the output is **n+2**; hence we predict the next values at the start and the end while reconstructing the input in the middle.

The ideas introduced in this model are:

1. **Masking anomaly scores**: Exponential weighted moving average function used to smooth the anomaly score produces unwanted false positives at the beginning. Masking the anomaly scores with the minimum of the anomaly scores reduces false positives.

2. **Bi-directional Regression**: Prediction-based models require at least _n_ observations before making the first prediction at the _n+1_ index. If _n_ is large, there are many false negatives at the start of the anomalies. If we perform regression in both directions (forward _f_ and reverse _r_), we can fill in the gap at the beginning of the anomaly scores. Overlapping scores in the middle are averaged. 
image

3. **Prediction-Reconstruction Mixture Errors**: We will combine the prediction and reconstruction scores in a similar way to the critic and reconstructions scores from TadGAN. The reconstruction scores default to DTW. The four methods to combine the two scores are multiplication, summation, prediction-only, and reconstruction-only.

### PR DESCRIPTION
**Pull request by lcwong0928**
Resolve #286 

**New**
- Add aer primitive, json files, and pipeline files

### CODE REVIEW

**Review by sarahmish — COMMENTED** [review]
Looks great @lcwong0928.

As a last step, can you please add documentation in the respective files:
1. **pipeline documentation**. Add the name of the pipeline `AER` to the [table](https://github.com/sintel-dev/Orion/blob/master/docs/user_guides/primitives_pipelines/pipelines.rst#current-available-pipelines) as well as a one line description next to it. I would prefer if you add it after `Dense AE`.

2. **primitive documentation**. Add a primitive description to `aer` primitive similar to tadgan primitive [documentation](https://github.com/sintel-dev/Orion/blob/master/docs/user_guides/primitives_pipelines/primitives/TadGAN.rst). Then add it to the index tree of the documentation [here](https://github.com/sintel-dev/Orion/blob/master/docs/user_guides/primitives_pipelines/primitives.rst).

File: orion/pipelines/verified/aer/aer_realtraffic.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "mlprimitives.custom.timeseries_preprocessing.time_segments_aggregate#1": {
+        "time_column": "timestamp",
+        "interval": 600,
+        "method": "mean"
+    },
+    "orion.primitives.aer.score_anomalies#1": {
+        "comb": "reg"
```
[the remark this answers is not in the document]
Comment (lcwong0928) [reply orion/pipelines/verified/aer/aer_realtraffic.json:8]: The parameter is chosen in an ablation study similar to TadGAN's ablation study. The combination (reg, rec, sum, mult) with the highest F1 score for the dataset is chosen.

File: orion/primitives/aer.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+        """Fit the model.
+
+        Args:
+            X (ndarray):
+                N-dimensional array containing the input sequences for the model.
+            y (ndarray):
+                N-dimensional array containing the output sequences we want to reconstruct.
+        """
+        X = X[:, 1:-1, :]
+        ry, y, fy = y[:, 0], y[:, 1:-1], y[:, -1]
```
[the remark this answers is not in the document]
Comment (lcwong0928) [reply orion/primitives/aer.py:157]: Added case.

File: orion/primitives/aer.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                N-dimensional array containing the regression for each input sequence (reverse).
+        """
+        X = X[:, 1:-1, :]
+        ry, y, fy = self.aer_model.predict(X)
+        return ry, y, fy
+
+
+def bi_regression_errors(y: ndarray, ry_hat: ndarray, fy_hat: ndarray,
+                         smoothing_window: float = 0.01, smooth: bool = True, mask: bool = True):
+    """Compute an array of absolute errors comparing predictions and expected output.
```
[the remark this answers is not in the document]
Comment (lcwong0928) [reply orion/primitives/aer.py:198]: Updated docstrings

**Comment by lcwong0928** [comment]
@sarahmish would you want me to also update the benchmark file to include `aer` as well?

File: orion/primitives/aer.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        smoothing_window (float):
+            Optional. Size of the smoothing window, expressed as a proportion of the total
+            length of y. If not given, 0.01 is used.
+        smooth (bool):
+            Optional. Indicates whether the returned errors should be smoothed with EWMA.
+            If not given, `True` is used.
+        mask (bool): bool = True
+            Optional. Mask anomaly score errors in the beginning.
+            If not given, `True` is used.
+    Returns:
```
Comment (sarahmish) [inline orion/primitives/aer.py:226]: minor: add space between the last a item in argument and returns.

File: orion/primitives/aer.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        rec_error_type (str):
+            Optional. The method to compute reconstruction error. Can be one of
+            `["point", "area", "dtw"]`. If not given, 'dtw' is used.
+        comb (str):
+            Optional. How to combine critic and reconstruction error. Can be one
+            of `["mult", "sum", "rec"]`. If not given, 'mult' is used.
+        lambda_rec (float):
+            Optional. Used if `comb="sum"` as a lambda weighted sum to combine
+            scores. If not given, 0.5 is used.
+    Returns:
```
Comment (sarahmish) [inline orion/primitives/aer.py:281]: minor: add space between the last a item in argument and returns.

File: orion/primitives/aer.py (hunk 4 above)
Comment (lcwong0928) [reply orion/primitives/aer.py:226]: Added

File: orion/primitives/aer.py (hunk 5 above)
Comment (lcwong0928) [reply orion/primitives/aer.py:281]: Added

**Review by sarahmish — APPROVED** [review]
Perfect!

File: docs/user_guides/primitives_pipelines/primitives/AER.rst [hunk 6]
```diff
@@ -0,0 +1,54 @@
+.. highlight:: shell
+
+TadGAN
```
Comment (dyuliu) [inline docs/user_guides/primitives_pipelines/primitives/AER.rst:3]: AER

File: docs/user_guides/primitives_pipelines/primitives/AER.rst [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ ``optimizer``              ``str``             string (name of optimizer) or optimizer instance. Default is ``keras.optimizers.Adam``
+ ``learning_rate``          ``float``           float denoting the learning rate of the optimizer. Default is 0.001
+ ``batch_size``             ``int``             number of samples per gradient update. Default is 64
+ ``layers_encoder``         ``list``            list containing layers of encoder
+ ``layers_generator``       ``list``            list containing layers of generator
+
+**output**
+------------------------------------------------------------------------------------------------------------------------------------------------
+
+ ``ry_hat``                 ``numpy.ndarray``    n-dimensional array containing the regression for each input sequence (forward)
```
Comment (dyuliu) [inline docs/user_guides/primitives_pipelines/primitives/AER.rst:33]: fy_hat

File: docs/user_guides/primitives_pipelines/primitives/AER.rst [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    import numpy as np
+    from mlprimitives import load_primitive
+
+    X = np.ones((64, 100, 1))
+    y = X[:,:, [0]] # signal to reconstruct from X (channel 0)
+    primitive = load_primitive('orion.primitives.aer.AER',
+        arguments={"X": X, "y": y, "epochs": 1, "batch_size": 1})
+
+    primitive.fit()
+    ry, y, fy = primitive.produce(X=X)
```
Comment (dyuliu) [inline docs/user_guides/primitives_pipelines/primitives/AER.rst:51]: Is this order correct? In the docstring, the order is forward, rec, reveve

File: orion/benchmark.py [hunk 9]
```diff
@@ -44,14 +44,18 @@
...
+    'aer', 'lstm_autoencoder', 'dense_autoencoder'
 ]
 
 VERIFIED_PIPELINES_GPU = {
     'arima': 'arima',
     'lstm_dynamic_threshold': 'lstm_dynamic_threshold_gpu',
     'azure': 'azure',
-    'tadgan': 'tadgan_gpu'
+    'tadgan': 'tadgan_gpu',
+    'aer': 'aer',
```
Comment (dyuliu) [inline orion/benchmark.py:56]: Can you confirm gpu version is the same with CPU version? The same question for `dense_autoencoder`

File: docs/user_guides/primitives_pipelines/primitives/AER.rst (hunk 6 above)
Comment (lcwong0928) [reply docs/user_guides/primitives_pipelines/primitives/AER.rst:3]: Updated

File: docs/user_guides/primitives_pipelines/primitives/AER.rst (hunk 7 above)
Comment (lcwong0928) [reply docs/user_guides/primitives_pipelines/primitives/AER.rst]: corrected typo

File: docs/user_guides/primitives_pipelines/primitives/AER.rst (hunk 8 above)
Comment (lcwong0928) [reply docs/user_guides/primitives_pipelines/primitives/AER.rst:51]: The order should be reverse, rec, forward (corrected in documentation)

File: orion/benchmark.py (hunk 9 above)
Comment (lcwong0928) [reply orion/benchmark.py:56]: The GPU version is the same for `dense_autoencoder` since it only uses the `dense` layer.

@sarahmish mentioned that I can ignore the `gpu` version of the `aer ` model. The pipeline file will be the same once we upgrade to TF2.0.
